kotlin-language-server: init at 1.1.1

critbase opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

critbase committed:
maintainers: add critbase

critbase committed:
kotlin-language-server: init at 1.1.1

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+{ lib, writeText, stdenv, fetchFromGitHub, jdk11, gradle, perl, kotlin, unzip
+, ... }:
```
```suggestion
{ lib, writeText, stdenv, fetchFromGitHub, jdk11, gradle, perl, kotlin, unzip }:
```

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+  '';
+
+  meta = with lib; {
+    description =
+      "Intelligent Kotlin support for any editor/IDE using the Language Server Protocol";
```
```suggestion
    description = "Intelligent Kotlin support for any editor/IDE using the Language Server Protocol";
```

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+  buildPhase = ''
+    export GRADLE_USER_HOME=$(mktemp -d);
+    gradle --no-daemon --offline --init-script ${gradleInit} -Porg.gradle.java.installations.auto-download=false -Dorg.gradle.java.home=${jdk11.home} :server:distZip
+  '';
+  installPhase = ''
```
```suggestion
  '';

  installPhase = ''
```

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+  '';
+  installPhase = ''
+    mkdir $out
+    ${unzip}/bin/unzip server/build/distributions/server.zip -d $out
```
```suggestion
    unzip server/build/distributions/server.zip -d $out
```

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+  src = fetchFromGitHub {
+    owner = "fwcd";
+    repo = "kotlin-language-server";
+    rev = "${version}";
+    sha256 = "4z6rpu3FYQRLNYeDo3NvTGTrV2/CMzUvQ+X3ZGNqszc=";
+  };
+  nativeBuildInputs = [ jdk11 gradle kotlin ];
```
```suggestion
    rev = version;
    sha256 = "4z6rpu3FYQRLNYeDo3NvTGTrV2/CMzUvQ+X3ZGNqszc=";
  };

  nativeBuildInputs = [ jdk11 gradle kotlin unzip ];
```

critbase commented:
@SuperSandro2000 i think i've applied all your suggestions, is there anything else i should do for this pr?

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+  '';
+in stdenv.mkDerivation rec {
+  pname = "kotlin-language-server";
+  version = "1.1.1";
```
```suggestion
  inherit version;
```

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+{ lib, writeText, stdenv, fetchFromGitHub, jdk11, gradle, perl, kotlin, unzip }:
+
+let
```
```suggestion
let
  version = "1.1.1";
```

SuperSandro2000 commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,76 @@
+    src = fetchFromGitHub {
+      owner = "fwcd";
+      repo = "kotlin-language-server";
+      rev = "1.1.1";
```
```suggestion
      rev = version;
```

SuperSandro2000 reviewed (changes requested):
just noticed that version shouldn't be defined twice.

critbase commented:
alright, done.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/534

futile commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,77 @@
+in stdenv.mkDerivation rec {
+  pname = "kotlin-language-server";
+  inherit version;
+  src = fetchFromGitHub {
+    owner = "fwcd";
+    repo = "kotlin-language-server";
+    rev = version;
+    sha256 = "4z6rpu3FYQRLNYeDo3NvTGTrV2/CMzUvQ+X3ZGNqszc=";
+  };
```
This seems to be the same as `deps.src`, maybe pull this out into its own binding?

critbase replied:
oh, good catch. thanks, doing this now.

futile commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,77 @@
+  buildInputs = [ jdk11 kotlin ];
+  buildPhase = ''
+    export GRADLE_USER_HOME=$(mktemp -d);
+    gradle --no-daemon --offline --init-script ${gradleInit} -Porg.gradle.java.installations.auto-download=false -Dorg.gradle.java.home=${jdk11.home} :server:distZip
```
This command line also seems to contain overlapping elements with the one from `deps.buildPhase`, maybe it also makes sense to extract these into a common binding? Just to reduce repetition/room for errors here. But might not be worth it.

critbase replied:
I think it's similar enough. I'll go ahead and do this.

futile commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,77 @@
+  installPhase = ''
+    mkdir $out
+    unzip server/build/distributions/server.zip -d $out
+    mv $out/server/* $out
```
Just wondering, might it be necessary here to adjust permissions on files using something like `install`? Or do these already have correct permissions for some reason?

critbase replied:
`install` is a bit more verbose than this, but that's not bad. I'll do this too, i like the idea.

futile commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,77 @@
+    homepage = "https://github.com/fwcd/kotlin-language-server";
+    license = licenses.mit;
+    maintainers = [ maintainers.critbase ];
+    platforms = platforms.linux;
```
Mhhm, maybe more platforms are supported? I'd think `darwin` might just be supported, and theoretically maybe even windows (even though nix itself doesn't run on windows yet)? So maybe `platforms.unix` or `platforms.all` might be possible here.

critbase replied:
I haven't tested on darwin, but it should (probably) work just the same. As for windows, there's a .bat file included that's deleted, currently. I'll go ahead and leave it in, just in case.

futile reviewed (commented):
Just gave it a quick glance and it seems fine in general, left only a few comments. Nice package! :)

critbase commented:
@futile I added most of your suggestions. I also fixed an issue I overlooked when the JDK/Java isn't available in `PATH`; `kotlin-language-server` errors out without it.

futile commented on pkgs/development/tools/kotlin-language-server/default.nix:
```diff
@@ -0,0 +1,83 @@
+
+    install buildoutput/server/bin/* $out/bin
+    install buildoutput/server/lib/* $out/lib
+    install buildoutput/server/licenseReport.html $out/share
```
Have you checked if the `install`ed files have correct permissions? Otherwise it should probably be easy to adjust them here.

futile commented:
> @futile I added most of your suggestions. I also fixed an issue I overlooked when the JDK/Java isn't available in `PATH`; `kotlin-language-server` errors out without it.

Sounds good! I'm not really familiar with Java-packaging in nixpkgs, so I can't really say whether this a recommended/usual solution. Is it also done like this in other packages/how do other packages solve this? That should give a good indication.

critbase commented:
I've decided to give up on this package. I only recently noticed that the built NixOS package is broken when trying to actually use it, for example with Emacs and lsp-mode. I haven't been able to figure this out, and honestly, I think it's beyond my skill as of now. Because of this, I don't think it's worth it for me to continue trying for now.
Maybe sometime in the future I'll try again, or when NixOS gains some tooling for handling Gradle a bit better.

pschyska commented:
@critbase I am using this package, and it seems to work fine as far as I can tell (with neovim's integrated lsp, configuration below). Do you have more information what failed in emacs' lsp-mode? I think it's worth trying to fix them, and I'd be willing to pick your PR up if you don't want to keep working on it. Can you give me some elisp snippets by any chance that minimally reproduces this? I'm not an emacs user and would have to find out how to configure lsp-mode myself.

Relevant neovim lsp config that I am using:
```lua
lspconfig.kotlin_language_server.setup {
  on_attach = on_attach,
  root_dir = function(fname) return lspconfig.util.find_git_ancestor(fname) end,
  settings = {kotlin = {compiler = {jvm = {target = "11"}}}}
}
```

Where on_attach is basically what [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig#keybindings-and-completion) recommended.

critbase commented:
@pschyska By all means, please go ahead. I've pretty much given up on this, and just use IntelliJ/android-studio for kotlin now. As for emacs, I consistently get an odd error that looks like this:
```
Command "kotlin-language-server" is present on the path.
Found the following clients for /home/miyuki/.bin/kotlin-quick-start/src/main/kotlin/quick/start/App.kt: (server-id kotlin-ls, priority -1)
The following clients were selected based on priority: (server-id kotlin-ls, priority -1)
main      Adding script definitions [ScriptTemplateWithArgs]
main      Kotlin Language Server: Version 1.1.1
main      Connected to client
async0    Internal error: java.lang.NullPointerException: folders must not be null
java.util.concurrent.CompletionException: java.lang.NullPointerException: folders must not be null
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException: folders must not be null
	at org.javacs.kt.KotlinLanguageServer$initialize$1.invoke(KotlinLanguageServer.kt:98)
	at org.javacs.kt.KotlinLanguageServer$initialize$1.invoke(KotlinLanguageServer.kt:22)
	at org.javacs.kt.util.AsyncExecutorKt$sam$java_util_function_Supplier$0.get(AsyncExecutor.kt)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 3 more
```
I'm not at all a java developer, and my use case for kotlin is only developing android apps, so I decided to give up on this and just switch to android studio.

pschyska commented:
Just a heads-up, I also gave up on kotlin-language-server for now. It seems impossible to "go to definition" into dependencies, which makes it too awkward to effectively work on kotlin code for me. As (in my case) most dependencies are in Java, it would require collaboration with a java language server, which is something that upstream has to implement AFAICS (related https://github.com/fwcd/kotlin-language-server/issues/247.
